test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperDescriptorProviderTest.cs

```diff
@@ -1,16 +1,105 @@
 ﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Collections.Generic;
 using System.Linq;
-using Microsoft.AspNet.Razor.TagHelpers;
 using Xunit;
 
-namespace Microsoft.AspNet.Razor.Test.TagHelpers
+namespace Microsoft.AspNet.Razor.TagHelpers
 {
     public class TagHelperDescriptorProviderTest
     {
+        public static TheoryData RequiredAttributeData
+        {
+            get
+            {
+                var divDescriptor = new TagHelperDescriptor(
+                    tagName: "div",
+                    typeName: "DivTagHelper",
+                    assemblyName: "SomeAssembly",
+                    attributes: Enumerable.Empty<TagHelperAttributeDescriptor>(),
+                    requiredAttributes: new[] { "style" });
+                var inputDescriptor = new TagHelperDescriptor(
+                    tagName: "input",
+                    typeName: "InputTagHelper",
+                    assemblyName: "SomeAssembly",
+                    attributes: Enumerable.Empty<TagHelperAttributeDescriptor>(),
+                    requiredAttributes: new[] { "class", "style" });
+                var catchAllDescriptor = new TagHelperDescriptor(
+                    tagName: "*",
+                    typeName: "CatchAllTagHelper",
+                    assemblyName: "SomeAssembly",
+                    attributes: Enumerable.Empty<TagHelperAttributeDescriptor>(),
+                    requiredAttributes: new[] { "class" });
+                var catchAllDescriptor2 = new TagHelperDescriptor(
+                    tagName: "*",
+                    typeName: "CatchAllTagHelper",
+                    assemblyName: "SomeAssembly",
+                    attributes: Enumerable.Empty<TagHelperAttributeDescriptor>(),
+                    requiredAttributes: new[] { "custom", "class" });
+                var defaultAvailableDescriptors =
+                    new[] { divDescriptor, inputDescriptor, catchAllDescriptor, catchAllDescriptor2 };
+
+                return new TheoryData<
+                    string, // tagName
+                    IEnumerable<string>, // providedAttributes
+                    IEnumerable<TagHelperDescriptor>, // expectedDescriptors
+                    IEnumerable<TagHelperDescriptor>> // availableDescriptors
+                {
+                    { "div", new[] { "custom" }, new TagHelperDescriptor[0],  defaultAvailableDescriptors },
```

Comment:
curious: why the random switches between `Enumerable.Empty<T>()` and `new T[0]`?  should probably stick with `Enumerable.Empty<T>()` but this is test code so it's not a requirement.